source/extensions/filters/network/mysql_proxy/mysql_utils.cc

```diff
@@ -216,6 +226,129 @@ DecodeStatus BufferHelper::peekHdr(Buffer::Instance& buffer, uint32_t& len, uint
+     */
+    bool v41 = cap & CLIENT_PROTOCOL_41;
+    bool sconn = cap & CLIENT_SECURE_CONNECTION;
+    if (!v41 || !sconn) {
+      return AuthMethod::OldPassword;
+    }
+    return AuthMethod::NativePassword;
+  }
+
+  if (auth_plugin_name == MYSQL_OLD_PASSWORD) {
+    return AuthMethod::OldPassword;
+  }
+  if (auth_plugin_name == MYSQL_NATIVE_PASSWORD) {
+    return AuthMethod::NativePassword;
+  }
+  if (auth_plugin_name == MYSQL_SHA256_PASSWORD) {
+    return AuthMethod::Sha256Password;
+  }
+  if (auth_plugin_name == MYSQL_CACHE_SHA_PASSWORD) {
+    return AuthMethod::CacheSha2Password;
+  }
+  if (auth_plugin_name == MYSQL_CLEAR_PASSWORD) {
+    return AuthMethod::ClearPassword;
+  }
+  return AuthMethod::Unknown;
+}
+
+// https://github.com/mysql/mysql-server/blob/5.5/sql/password.c#L186
+std::vector<uint8_t> OldPassword::signature(const std::string& password,
+                                            const std::vector<uint8_t>& seed) {
+  std::vector<uint8_t> to;
+  if (password.empty()) {
+    return to;
+  }
+  to.resize(SEED_LENGTH);
+  auto hash_pass = hash(password);
+  auto hash_seed = hash(seed);
+  RandStruct rand_st(hash_pass[0] ^ hash_seed[0], hash_pass[1] ^ hash_seed[1]);
+  for (int i = 0; i < SEED_LENGTH; i++) {
+    to[i] = static_cast<uint8_t>((floor(rand_st.myRnd() * 31) + 64));
+  }
+  char extra = static_cast<char>(floor(rand_st.myRnd() * 31));
+  for (int i = 0; i < SEED_LENGTH; i++) {
+    to[i] ^= extra;
+  }
+  return to;
+}
+
+std::vector<uint8_t> NativePassword::generateSeed() {
+  std::vector<uint8_t> res(SEED_LENGTH);
+  RAND_bytes(res.data(), SEED_LENGTH);
+  return res;
+}
+
+std::vector<uint8_t> NativePassword::signature(const std::string& password,
+                                               const std::vector<uint8_t>& seed) {
+  auto hashstage1 = hash(password);
+  auto hashstage2 = hash(hashstage1);
+
+  auto text = seed;
+  text.insert(text.end(), hashstage2.begin(), hashstage2.end());
+  auto to_be_xor = hash(text);
+
+  for (int i = 0; i < SHA_DIGEST_LENGTH; i++) {
+    to_be_xor[i] = to_be_xor[i] ^ hashstage1[i];
+  }
+  return to_be_xor;
+}
+
+std::vector<uint8_t> NativePassword::hash(const char* text, int size) {
+  bssl::ScopedEVP_MD_CTX ctx;
+  std::vector<uint8_t> result(SHA_DIGEST_LENGTH);
+  auto rc = EVP_DigestInit(ctx.get(), EVP_sha1());
+  RELEASE_ASSERT(rc == 1, "Failed to init digest context");
+  rc = EVP_DigestUpdate(ctx.get(), text, size);
+  RELEASE_ASSERT(rc == 1, "Failed to update digest");
+  rc = EVP_DigestFinal(ctx.get(), result.data(), nullptr);
+  RELEASE_ASSERT(rc == 1, "Failed to finalize digest");
+  return result;
+}
+
+std::vector<uint8_t> OldPassword::generateSeed() {
+  std::vector<uint8_t> res(SEED_LENGTH);
+  RAND_bytes(res.data(), SEED_LENGTH);
+  return res;
+}
+
+std::vector<uint32_t> OldPassword::hash(const char* text, int size) {
+  uint32_t nr = 1345345333L, add = 7, nr2 = 0x12345671L;
```

Comment:
Can you provide the link to documentation explaining the meaning of those magic numbers?

Reply:
add paper ref.